torch/csrc/distributed/c10d/init.cpp

```diff
@@ -301,6 +301,11 @@ to apply layer-wise operations.
 Returns:
     A list of ``torch.Tensor``. Each tensor in the list corresponds to a parameter.
 )")
+      .def(
+          "get_grad_index_to_variable_mapping",
```

Comment:
curious, is the param index sth that users must know? Will it work if we just expose one single API `bucket.parameters()` that returns references to model parameters with the `.grad` field properly populated?

Reply:
Above might only work when `grad_as_bucket_view` is enabled. Curious, when `grad_as_bucket_view=False`, how will optimizer work in the communication hook? Are we expecting users to manually populate `param.grad` before calling `optimizer.step()`? If so, will it work if we expose `bucket.parameters` and `bucket.grads` as two separate APIs, where parameters() return references to model parameters and `grads()` return bucket views?

Reply:
Regarding bucket.parameters and then using param.grad field, that might be hard since we will be running optimizer as a `then()` callback to gradient reduction, but `param.grad` is not populated until `Reducer::finalize_backward` is called. So if we access `param.grad` it will be a stale gradient in the comm. hook.

> Curious, when grad_as_bucket_view=False, how will optimizer work in the communication hook? Are we expecting users to manually populate param.grad before calling optimizer.step()? 

In both cases, we are planning to do something like the following:
```
grad_tensors = bucket.get_per_parameter_tensors()
idx_to_param = bucket.get_idx_to_params()
for i in range(len(grad_tensors)):
    grad, model_param = grad_tensors[i], idx_to_param[i]
    functional_optimizer.step(model_param, grad)
```

For this to work we'll need functional optimizers that operate on explicit gradients which we've created a version of for the RPC framework.

Reply:
Will sth like this work:

```python
for param, grad in zip(bucket.parameters(), bucket.gradients()):
    # param is the model parameter, whose .grad field is not populated yet
    # grad is a view to the bucket offset, and it already contains the allreduced grad. 
    functional_optimizer.step(model_param, grad)
```

Reply:
Above code means that `parameters()` and `gradients()` must match in order, but it should be fine as they both belong to the same bucket. And it will also avoid engaging with param index.

Reply:
Probably this can be a follow-up refactoring PR.